File: rusty_web/src/api/pipelines.rs
use commons::errors::RustyError;
use domain::pipelines::{PagedPipelines, Pipeline, RegisterPipeline};

use crate::api::client::reqwasm_post;
use crate::api::utils::{parse_entries, parse_paged};

/// Function to retrieve pipelines for job from a GraphQL endpoint.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn get_pipelines_for_job(job_id: String) -> Result<PagedPipelines, RustyError> {
    let payload = serde_json::json!({
        "query": format!(r#"query {{
            pipelines {{
                get(
                    filter: {{ job_id: "{}" }},
        			options: {{ sortMode: DESCENDING, sortField: "number", pageSize: 99 }}
                ) {{
                    total
                    page
                    pageSize
                    entries {{
                        id
                        number
                        startDate
                        registerDate
                        endDate
                        status
                        jobId
                        agentId
                    }}
                }}
            }}
        }}"#, job_id),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: serde_json::Value = serde_json::from_str(&data)?;
    let json_data = json_data["data"]["pipelines"]["get"].clone();
    let (total, page, page_size, entries) = parse_paged(&json_data)?;
    let entries = parse_entries(entries)?;
    Ok(PagedPipelines {
        total,
        page,
        page_size,
        entries,
    })
}

/// Function to retrieve last pipeline for job from a GraphQL endpoint.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn get_last_pipeline_for_job(job_id: String) -> Result<Option<Pipeline>, RustyError> {
    let payload = serde_json::json!({
        "query": format!(r#"query {{
            pipelines {{
                get(filter: {{
                    job_id: "{}"
                }}, options: {{
                    sortMode: "DESCENDING",
                    sortField: "number",
                    pageSize: 1
                }}) {{
                    total
                    page
                    pageSize
                    entries {{
                        id
                        number
                        startDate
                        registerDate
                        endDate
                        status
                        jobId
                        agentId
                    }}
                }}
            }}
        }}"#, job_id),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: serde_json::Value = serde_json::from_str(&data)?;
    let json_data = json_data["data"]["pipelines"]["get"].clone();
    let (_, _, _, entries) = parse_paged(&json_data)?;
    let entries: Vec<Pipeline> = parse_entries(entries)?;
    Ok(if entries.len() == 1 {
        entries.first().cloned()
    } else {
        None
    })
}

/// Function to run a new pipeline via GraphQL endpoint.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn run_pipeline(model: RegisterPipeline) -> Result<String, RustyError> {
    let payload = serde_json::json!({
        "query": format!(r#"mutation {{
            pipelines {{
                register(pipeline: {{
                    jobId: "{}"
                }})
            }}
        }}"#, model.job_id),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: serde_json::Value = serde_json::from_str(&data)?;
    Ok(json_data["data"]["pipelines"]["register"].to_string())
}

File: rusty_web/src/lib/api/pipelines.ts
import { fetchPost } from '$lib/utils/api';
import type { Pipeline } from '$lib/domain/pipeline';

const getJobPipelinesQuery = (page: number, id: string) => {
	const filter = `filter: { job_id: { equals: "${id}" } }, `;
	const options = `options: { pageNumber: ${page}, pageSize: 20, sortMode: DESCENDING, sortField: "number" }`;

	return `query {
		pipelines {
			get(${filter}${options}){
				total
				page
				pageSize
				entries {
					id
					number
					status
					stageStatus
					branch
					registerDate
					startDate
					endDate
					jobId
				}
			}
		}
	}`;
};

export const getJobPipelines = async (auth: string, page: number, id: string) => {
	try {
		const response = await fetchPost(
			auth,
			JSON.stringify({ query: getJobPipelinesQuery(page, id) })
		);

		if (!response.ok) {
			return {
				errors: ['Fetch job pipelines failed']
			};
		} else {
			const { data, errors } = await response.json();
			if (errors && errors.length > 0) {
				return {
					errors: errors.map((error: { message: string }) => error.message)
				};
			} else if (data) {
				const paged = data?.pipelines?.get;
				const pipelines: Pipeline[] = paged?.entries ?? [];
				return {
					total: paged?.total ?? 0,
					page: paged?.page ?? 1,
					pageSize: paged?.pageSize ?? 20,
					entries: pipelines
				};
			}
		}
	} catch (error) {
		return {
			errors: ['Fetch job pipelines failed']
		};
	}
};

const registerPipelineMutation = (jobId: string, branch: string) => {
	return `mutation {
		pipelines {
			register(pipeline: {
				jobId: "${jobId}",
				branch: "${branch}"
			})
		}
	}`;
};

export const registerPipeline = async (auth: string, jobId: string, branch: string) => {
	try {
		const response = await fetchPost(
			auth,
			JSON.stringify({ query: registerPipelineMutation(jobId, branch) })
		);

		if (!response.ok) {
			return {
				errors: ['Register pipeline failed']
			};
		} else {
			const { data, errors } = await response.json();
			if (errors && errors.length > 0) {
				return {
					errors: errors.map((error: { message: string }) => error.message)
				};
			} else if (data) {
				return data;
			}
		}
	} catch (error) {
		return {
			errors: ['Register pipeline failed']
		};
	}
};

const getPipelineByIdQuery = (id: string) => {
	return `query {
		pipelines {
			getById(id: "${id}"){
                id
                number
                branch
                status
				stageStatus
                registerDate
                startDate
                endDate
                jobId
                agentId
			}
		}
	}`;
};

export const getPipelineById = async (auth: string, id: string) => {
	try {
		const response = await fetchPost(auth, JSON.stringify({ query: getPipelineByIdQuery(id) }));

		if (!response.ok) {
			return {
				errors: ['Get pipeline by id failed']
			};
		} else {
			const { data, errors } = await response.json();
			if (errors && errors.length > 0) {
				return {
					errors: errors.map((error: { message: string }) => error.message)
				};
			} else if (data) {
				return data?.pipelines?.getById;
			}
		}
	} catch (error) {
		return {
			errors: ['Get pipeline by id failed']
		};
	}
};

const getPipelineByLogsQuery = (id: string) => {
	return `query { pipelines { getLogs(id: "${id}") } }`;
};

export const getPipelineLogs = async (auth: string, id: string) => {
	try {
		const response = await fetchPost(auth, JSON.stringify({ query: getPipelineByLogsQuery(id) }));

		if (!response.ok) {
			return {
				errors: ['Get pipeline logs failed']
			};
		} else {
			const { data, errors } = await response.json();
			if (errors && errors.length > 0) {
				return {
					errors: errors.map((error: { message: string }) => error.message)
				};
			} else if (data) {
				return data?.pipelines?.getLogs
					.map((value: string) => JSON.parse(value))
					.reduce((rv: any, x: any) => {
						let key = x['stage'];
						if (key == 'rusty-before') {
							key = 'before';
						}
						if (key == 'rusty-after') {
							key = 'after';
						}
						(rv[key] = rv[key] || []).push(x);
						return rv;
					}, {});
			}
		}
	} catch (error) {
		return {
			errors: ['Get pipeline logs failed']
		};
	}
};
